src/IO/Observer/Actions.hpp

```diff
@@ -32,11 +32,12 @@ namespace Actions {
 /// Should be invoked on ObserverWriter.
 struct RegisterVolumeContributorWithObserverWriter {
  public:
-  template <typename... DbTags, typename... InboxTags, typename Metavariables,
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
             typename ArrayIndex, typename ActionList,
             typename ParallelComponent, typename... ReductionDatums,
-            Requires<sizeof...(DbTags) != 0> = nullptr>
-  static void apply(db::DataBox<tmpl::list<DbTags...>>& box,
+            Requires<tmpl::list_contains_v<
+                DbTagsList, Tags::VolumeObserversRegistered>> = nullptr>
```

Comment:
This is going to be terrible for compile times. I saw your implementation on your branch and intentionally didn't use this because of that reason. The only place this could be an issue is the first action in the phase and the initialization code. Even in those cases I expect that the vast majority of the time checking one is sufficient since it's very unlikely another action only adds `A` when the one you want the fancier constraint on adds `A` and `B`.

Reply:
Yes, I'm not suggesting to check for _all_ dependencies or added tags, but just for a few for the SFINAE resolution. Further down in this file you chain multiple `list_contains_v ` together and I think those can be replaced by a `db::can_retrieve_all` utility, with the bonus of supporting compute items.